hudi-common/src/main/java/org/apache/hudi/common/util/collection/ExternalSpillableMap.java

```diff
@@ -202,22 +199,19 @@ public R get(Object key) {
 
   @Override
   public R put(T key, R value) {
+    if (this.currentInMemoryMapSize >= maxInMemorySizeInBytes || inMemoryMap.size() % NUMBER_OF_RECORDS_TO_ESTIMATE_PAYLOAD_SIZE == 0) {
```

Comment:
I am not sure how this refactoring results in more accurate estimation of size. can you help me understand please.

Reply:
I put the real use case in the JIRA description ([link](https://issues.apache.org/jira/browse/HUDI-4753)). 

In short, the original implementation compute `avgSize` based on the first record, which may be very large in certain case, i.e., metadata table. Causing unbalanced log block which in turn hurt write&read performance.

Reply:
makes sense.

Reply:
What is the purpose for estimation for check: `this.currentInMemoryMapSize >= maxInMemorySizeInBytes` ?


And why we have this evaluate expression:
```java
this.estimatedPayloadSize = (long) (this.estimatedPayloadSize * 0.9 
        + (keySizeEstimator.sizeEstimate(key) + valueSizeEstimator.sizeEstimate(value)) * 0.1)
```